### ISSUE TITLE
View functions return inconsistent view of weights

### ISSUE THREAD

**Issue by Trisome-Bedrock** [Bedrock-Technology/bedrock-dao#44]
- [x] Problem 1:

According to local testing, [getTotalWeight](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L413) returns 300000000000000000000, while [gaugeRelativeWeight](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L353) returns 0. 
```
got_total_weight:  300000000000000000000
got_gauge_relative_weight_2:  0
got_gauge_relative_weight_3:  0

Fast forwad the lock by one week

got_total_weight:  300000000000000000000
got_gauge_relative_weight_2:  333333333333333333
got_gauge_relative_weight_3:  666666666666666666
```

- [x] Problem 2:

The function [getUserVotesWtForGauge](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L392) reverted because the gauge base weight had not been accounted for in the [gaugePoints](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L717) map at the given timestamp.

- [x] Problem 3:

The function [getGaugeWeight](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L367) may return an outdated checkpointed value, which is not guaranteed to be always updated.

- [x] Problem 4: 
The function [_getGaugeWeightReadOnly](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L707) disregards base weight and slope variations.

- [x] Problem 5:
The function [_gaugeRelativeWeight](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L741) could return zero if there exists a checkpoint gap for the historical data.

- [x] Problem 6:
The function [getTypeWeight](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L405) could panic due to integer overflow.

- [x] Problem 7:
The function [getTotalWeight](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L413) could return wrong value for current week if there exists a checkpoint gap for the historical data.

- [x] Problem 8:
The function [getWeightsSumPerType](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L422) could return wrong value for current week if there exists a checkpoint gap for the historical data. Besides, it could also panic due to integer overflow.

- [x] Problem 9:
The function [getUserVotesWtForGauge](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L392) could return an incorrect value because it calculates user votes using the latest gauge base weight, not the corresponding base weight at the time specified by the user.

- [x] Problem 10:
The issue [#28](https://github.com/Bedrock-Technology/bedrock-dao/issues/28) results in inconsistent query weights.

**Comment by Trisome-Bedrock** [comment]
The function [getUserVotesWtForGauge](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L392) reverted because the gauge base weight had not been accounted for in the [gaugePoints](https://github.com/Bedrock-Technology/bedrock-dao/blob/11d53bdfff1ba05be059f9cda997b29cf26a344d/contracts/gauge_controller.sol#L717) map at the given timestamp.

**Comment by Trisome-Bedrock** [comment]
The code update for this issue will be merged and tested alongside issue #56.
